channeldb/migrations_test.go

```diff
@@ -5,15 +5,439 @@ import (
+					err)
+			}
+
+			err = chanIndex.Delete(op.Bytes())
+			if err != nil {
+				return fmt.Errorf("unable to delete from chan "+
+					"index: %v", err)
+			}
+		}
+
+		return nil
+	}
+
+	// helper method that asserts the policy with the original status is in
+	// the expected state after migration.
+	assertMigratedPolicy := func(tx *bbolt.Tx, node [33]byte,
+		channelID uint64, status policyStatus) error {
+
+		edges := tx.Bucket(edgeBucket)
+		if edges == nil {
+			return fmt.Errorf("edge bucket did not exist")
+		}
+
+		var edgeKey [33 + 8]byte
+		copy(edgeKey[:], node[:])
+		byteOrder.PutUint64(edgeKey[33:], channelID)
+
+		edgeBytes := edges.Get(edgeKey[:])
+
+		// Policy statuses should should be either complete or unknown.
+		switch status {
+
+		// Existing policy should be here still, and not unknown.
+		case existing:
+			if edgeBytes == nil || len(edgeBytes) < 10 {
+				return fmt.Errorf("expected existing policy " +
+					"to be present")
+			}
+
+		// Otherwise it should be unknown.
+		case nonExisting:
+			fallthrough
+		case unknown:
+			if !bytes.Equal(edgeBytes, unknownPolicy) {
+				return fmt.Errorf("expected policy to " +
+					"be unknown")
+			}
+		}
+		return nil
+	}
+
+	// Run through all test cases.
+	for _, test := range tests {
+		var chanID uint64
+
+		// beforeMigrationFunc will take the db into the state set by
+		// this particular test case.
+		beforeMigrationFunc := func(db *DB) {
+			graph := db.ChannelGraph()
+
+			// We being by adding the edge, the two nodes and the
+			// two channel edge policies to the DB.
+			info, policy1, policy2 := createTestEdge(t, db)
+			if err := graph.AddChannelEdge(info); err != nil {
+				t.Fatalf("unable to add edge: %v", err)
+			}
+			if err := graph.UpdateEdgePolicy(policy1); err != nil {
+				t.Fatalf("unable to update edge: %v", err)
+			}
+			if err := graph.UpdateEdgePolicy(policy2); err != nil {
+				t.Fatalf("unable to update edge: %v", err)
+			}
+
+			// Then we delete or modify the information in the
+			// database according to the testcase.
+			err := db.Update(func(tx *bbolt.Tx) error {
+
+				// Set the edge policies.
+				err := setPolicy(
+					tx, info.NodeKey1Bytes,
+					policy1.ChannelID, test.policy1Status,
+				)
+				if err != nil {
+					return err
+				}
+
+				err = setPolicy(
+					tx, info.NodeKey2Bytes,
+					policy2.ChannelID, test.policy2Status,
+				)
+				if err != nil {
+					return err
+				}
+
+				// Set the nodes.
+				err = setNode(
+					tx, info.NodeKey1Bytes,
+					test.node1Exists,
+				)
+				if err != nil {
+					return err
+				}
+
+				err = setNode(
+					tx, info.NodeKey2Bytes,
+					test.node2Exists,
+				)
+				if err != nil {
+					return err
+				}
+
+				// And finally set the edge info.
+				err = setInfo(tx, info, test.edgeInfoExists)
+				if err != nil {
+					return err
+				}
+
+				return nil
+			})
+			if err != nil {
+				t.Fatalf("unable to update db: %v", err)
+			}
```

Comment:
This is just the test though, do you still think we should add the data in a more low-level way?

Reply:
My rationale for doing it this way was that it made the code shorter to use the high level methods, then remove the specific data I wanted to test missing, than add all the non-missing data bit by bit.

Reply:
It is a more general question of how to test database migrations. To test a migration, you'd need to have a database with the old structure. In a test, this database needs to be set up. But setting that old db up with functions that assume the new structure already, is it robust?

Suppose one of those graph functions changes. It will impact this test too, but the test may still pass. Only at that point, it isn't testing the intended migration anymore.

I think it is better to setup the old database in a way that cannot change with future commits.